### PR TITLE
【Hackathon 6th Article No.6】稀疏计算的使用指南

### PR DESCRIPTION
**Pull request by KeithMaxwell**

### CODE REVIEW

File: rfcs/Article/guide_to_use_sparse.md [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+        out = sparse.add(out, identity)
+        out = self.relu(out)
+
+        return out
+```
+
+可见，ResNet 稀疏网络的代码和常规 ResNet 网络代码几乎没有差别。通过增加 import 路径替换，原网络代码基本都无需改动。通过 `from paddle.sparse import nn`，则可保持与原来的`nn.*`写法一致，更易于上手。
+
+## 3. 3D点云 CenterPoint
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:283]: 改成：Paddle 稀疏计算实战案例  吧

File: rfcs/Article/guide_to_use_sparse.md [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:284]: 介绍一下，将以 3D点云CenterPoint 为例来介绍

File: rfcs/Article/guide_to_use_sparse.md [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+4. CSR (Compressed Sparse Row)：CSR格式使用三个数组来表示稀疏矩阵。第一个数组存储非零元素的值，第二个数组存储每行的起始索引，第三个数组存储非零元素的列索引。
+5. CSC (Compressed Sparse Column)：CSC格式类似于CSR，但是以列为主的方式来存储数据。第一个数组存储非零元素的值，第二个数组存储每列的起始索引，第三个数组存储非零元素的行索引。
+6. DIA (Diagonal)：DIA格式是一种稀疏矩阵的存储格式，主要用于存储对角线上有值的稀疏矩阵。DIA格式使用两个数组来表示稀疏矩阵。第一个数组`data`存储非零元素的值，第二个数组`offsets`存储每个对角线的偏移量。
+
+这些格式各有优劣，选择取决于具体应用场景和算法需求。
+
+PaddlePaddle支持的主要稀疏格式包括：
+
+- COO格式（Coordinate Format）：用坐标表示非零元素的稀疏矩阵，包括三个数组：行索引、列索引和值。
+- CSR格式（Compressed Sparse Row）：将稀疏矩阵的行压缩，以节省存储空间。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:19]: 包含 几个数组 介绍下

File: rfcs/Article/guide_to_use_sparse.md [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+Paddle提供了与稠密计算高度一致的稀疏计算接口，易于上手。
+
+### 2.1 稀疏矩阵运算
+
+首先实现一个生成随机稀疏矩阵的函数：
+
+```python
+import numpy as np
+import paddle.sparse as sparse
+def random_sparse_tensor(shape, density, sparse_type='coo'):
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:139]: 这个可以写简单一些吗，直接paddle.rand 然后随机mask或dropout，然后在 to_sparse_coo/to_sparse_csr 就可以

**Comment by KeithMaxwell** [comment]
@zhwesky2010 已经全部按照要求修改了，麻烦研发老师再看看

File: rfcs/Article/guide_to_use_sparse.md [hunk 5]
```diff
@@ -15,15 +15,22 @@
 
 PaddlePaddle支持的主要稀疏格式包括：
 
-- COO格式（Coordinate Format）：用坐标表示非零元素的稀疏矩阵，包括三个数组：行索引、列索引和值。
-- CSR格式（Compressed Sparse Row）：将稀疏矩阵的行压缩，以节省存储空间。
+- COO格式（Coordinate Format）：此格式使用坐标来表示稀疏矩阵中的非零元素，涉及三个数组：行索引、列索引和值。
+- CSR格式（Compressed Sparse Row）：此格式通过压缩稀疏矩阵的行来节省存储空间。它包含三个数组，即Index Pointers，indices和Data数组。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:19]: 行指针信息、列坐标和值

File: rfcs/Article/guide_to_use_sparse.md [hunk 6]
```diff
@@ -15,15 +15,22 @@
 
 PaddlePaddle支持的主要稀疏格式包括：
 
-- COO格式（Coordinate Format）：用坐标表示非零元素的稀疏矩阵，包括三个数组：行索引、列索引和值。
-- CSR格式（Compressed Sparse Row）：将稀疏矩阵的行压缩，以节省存储空间。
+- COO格式（Coordinate Format）：此格式使用坐标来表示稀疏矩阵中的非零元素，涉及三个数组：行索引、列索引和值。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:18]: 行坐标、列坐标和值

File: rfcs/Article/guide_to_use_sparse.md [hunk 7]
```diff
@@ -84,6 +91,12 @@ CSR格式也存储三个数组，分别是Index Pointers，indices以及Data数
 
 例如，第一个index pointers对是`[0,2]`，那么这是表示稀疏矩阵第0行（0在Index Pointers数组中的索引是0）中元素的信息，并且表示第0行中共有两个非零元素。而使用`Indices[0,2]`可以获得这两个元素的列索引，使用`Data[0,2]`获取这两个元素具体的值。
 
+在`paddle.sparse`的CSR实现中，我们也使用了三个列表：
+
+- 一维列表`crows`对应于Index Pointers数组；
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:96]: 同上

File: rfcs/Article/guide_to_use_sparse.md [hunk 8]
```diff
@@ -15,15 +15,22 @@
...
 
 COO格式：
 
 ![COO Matrix](images/coo.gif)
 
 COO格式存储三个数组：行索引（Row）、列索引（Column）以及值（Data）数组。使用Data数组中的元素的索引分别去访问Row数组和Column数组就可以得到该元素在原来矩阵中的位置。
 
+在`paddle.sparse`的COO实现中，使用了两个列表：
+
+- 二维列表`indices`，包含行索引和列索引；
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:31]: 同上

File: rfcs/Article/guide_to_use_sparse.md [hunk 9]
```diff
@@ -134,31 +147,18 @@ Paddle提供了与稠密计算高度一致的稀疏计算接口，易于上手
...
-    
-    values = np.random.rand(num_nonzero_elements)
-    
-    indices = []
-    indices.append(row_indices.tolist())
-    indices.append(column_indices.tolist())
-
+    dense_tensor = paddle.randn(shape)
+    dropout = paddle.nn.Dropout(p=density)
+    dense_tensor = dropout(dense_tensor)
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:155]: paddle.nn.functional.dropout

File: rfcs/Article/guide_to_use_sparse.md [hunk 10]
```diff
@@ -134,31 +147,18 @@ Paddle提供了与稠密计算高度一致的稀疏计算接口，易于上手
...
     if sparse_type == 'coo':
-        return sparse.sparse_coo_tensor(indices, values.tolist(), shape)
+        sparse_tensor = dense_tensor.to_sparse_coo(sparse_dim=dense_tensor.dim())
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:157]: sparse_dim用默认的就可以吧

**Comment by KeithMaxwell** [comment]
@zhwesky2010 
已经按照要求修改了，麻烦研发老师再看看

File: rfcs/Article/guide_to_use_sparse.md [hunk 11]
```diff
@@ -15,29 +15,31 @@
 
 PaddlePaddle支持的主要稀疏格式包括：
 
-- COO格式（Coordinate Format）：此格式使用坐标来表示稀疏矩阵中的非零元素，涉及三个数组：行索引、列索引和值。
-- CSR格式（Compressed Sparse Row）：此格式通过压缩稀疏矩阵的行来节省存储空间。它包含三个数组，即Index Pointers，indices和Data数组。
+- COO格式（Coordinate Format）：此格式使用坐标来表示稀疏矩阵中的非零元素，涉及三个数组：行坐标、列坐标和值数组。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:18]: 包含三个数组

File: rfcs/Article/guide_to_use_sparse.md [hunk 12]
```diff
@@ -78,24 +80,28 @@ CSR格式：
...
 
-CSR格式也存储三个数组，分别是Index Pointers，indices以及Data数组。
+> 图片来源：https://matteding.github.io/2019/04/25/sparse-matrices/
+
+CSR格式也存储三个数组，分别是行指针信息（Index Pointers），列坐标（indices）以及值（Data）数组。
+
+* 行指针信息数组中相邻的两个元素，假设它们的坐标分别是r和r+1，值分别为a和b，那么可以确定：
 
-* Index Pointers数组中相邻的两个元素可以确定两个信息。
+  * 指针对中第一个元素的坐标r是稀疏矩阵行号。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:89]: 这一句可以删掉

File: rfcs/Article/guide_to_use_sparse.md [hunk 13]
```diff
@@ -78,24 +80,28 @@ CSR格式：
...
 
-  * 首先，它们在数组中的位置（索引对中第一个元素的索引）是行号。
+  * 它们的差（b-a）是矩阵第r行中非零元素的个数。
 
-  * 其次，这些值表示Indices数组的 [start: stop] 切片，它们的差是每行中非零元素的个数。使用指针查找索引以确定数据中每个元素的列。
+  * 若a不等于b，可以用a和b两个值构造一个切片[a:b]，那么使用该切片访问列坐标数组就可以得到第r行中非零元素的列坐标；若使用该切片访问值数组就可以得到第r行中非零元素的值。
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:93]: 写简单一点，若a不等于b，那么 indices[a: b]就是第r行中非零元素的列坐标、Data[a: b] 就是第r行中非零元素的值

File: rfcs/Article/guide_to_use_sparse.md [hunk 14]
```diff
@@ -78,24 +80,28 @@ CSR格式：
 
 ![](images/csr.gif)
 
-CSR格式也存储三个数组，分别是Index Pointers，indices以及Data数组。
+> 图片来源：https://matteding.github.io/2019/04/25/sparse-matrices/
+
+CSR格式也存储三个数组，分别是行指针信息（Index Pointers），列坐标（indices）以及值（Data）数组。
+
+* 行指针信息数组中相邻的两个元素，假设它们的坐标分别是r和r+1，值分别为a和b，那么可以确定：
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:87]: 行指针信息数组记录了....，假设其中两个相邻的元素坐标分别是...

File: rfcs/Article/guide_to_use_sparse.md [hunk 15]
```diff
@@ -78,24 +80,28 @@ CSR格式：
...
-  * 首先，它们在数组中的位置（索引对中第一个元素的索引）是行号。
+  * 它们的差（b-a）是矩阵第r行中非零元素的个数。
 
-  * 其次，这些值表示Indices数组的 [start: stop] 切片，它们的差是每行中非零元素的个数。使用指针查找索引以确定数据中每个元素的列。
+  * 若a不等于b，可以用a和b两个值构造一个切片[a:b]，那么使用该切片访问列坐标数组就可以得到第r行中非零元素的列坐标；若使用该切片访问值数组就可以得到第r行中非零元素的值。
 
-* Indices数组记录了每个元素的列索引。
-* Data数组记录了元素的值
+* 列坐标数组记录了非零元素的列坐标。
+* 值数组记录了非零元素的值
```
Comment (zhwesky2010) [inline rfcs/Article/guide_to_use_sparse.md:97]: 这个例子就不需要了，上面已经通过r、r+1举例

**Comment by KeithMaxwell** [comment]
@zhwesky2010  已经按照要求修改了，麻烦研发老师再看看

**Review by zhwesky2010 — APPROVED** [review]
LGTM
